api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,616 @@
+		func(p []*scaler.ResourceProvider, c []scaler.ResourceConsumer) {
+			sortProvidersByCPU(p)
+			sortProvidersByUsage(p)
+			sortConsumersByCPU(c)
+		},
+		func(p []*scaler.ResourceProvider, c []scaler.ResourceConsumer) {
+			sortProvidersByMemory(p)
+			sortProvidersByUsage(p)
+			sortConsumersByMemory(c)
+		},
+	}
+
+	for _, sort := range sorters {
+
+		// Because *ResourceProvider.SubtractResourcesFor() actually modifies the underlying
+		// provider object, we need to make a fresh copy of the list of providers for each
+		// pass through a sort-and-allocate operation.
+		resourceProvidersCopy := make([]*scaler.ResourceProvider, len(resourceProviders))
+		for i, provider := range resourceProviders {
+			resourceProvidersCopy[i] = scaler.NewResourceProvider(
+				&provider.AgentIsConnected,
+				provider.AvailableCPU,
+				provider.AvailableMemory,
+				provider.ID,
+				&provider.InUse,
+				&provider.Status,
+				&provider.UsedPorts,
+			)
+		}
+
+		if len(resourceProvidersCopy) == 0 && len(resourceConsumers) > 0 {
+			newProvider, err := e.calculateNewProvider(clusterName)
+			if err != nil {
+				return nil, []error{err}
+			}
+
+			newProvider.AgentIsConnected = true
+			newProvider.Status = ecs.ContainerInstanceStatusActive
+
+			resourceProvidersCopy = append(resourceProvidersCopy, newProvider)
+		}
+
+		providerDistribution := ProviderDistribution{}
+		providerDistribution.Providers = resourceProvidersCopy
+
+		providerDistributions = append(providerDistributions, providerDistribution)
+
+		sort(providerDistribution.Providers, resourceConsumers)
+
+		for _, consumer := range resourceConsumers {
+			hasRoom := false
+
+			for _, provider := range providerDistribution.Providers {
+				if provider.HasResourcesFor(consumer) {
+					hasRoom = true
+					if err := provider.SubtractResourcesFor(consumer); err != nil {
+						return nil, []error{err}
+					}
+					break
+				}
+			}
+
+			if !hasRoom {
+				newProvider, err := e.calculateNewProvider(clusterName)
+				if err != nil {
+					return nil, []error{err}
+				}
+
+				newProvider.AgentIsConnected = true
+				newProvider.Status = ecs.ContainerInstanceStatusActive
+
+				if !newProvider.HasResourcesFor(consumer) {
+					text := fmt.Sprintf("Resource '%s' cannot fit into an empty provider!", consumer.ID)
+					text += "\nThe instance size in your environment is too small to run this resource."
+					text += "\n Please increase the instance size for your environment or remove this resource."
+					err := errors.Newf(errors.IncompatibleConsumerAndProvider, text)
+					providerDistribution.Errs = append(providerDistribution.Errs, err)
+					continue
+				}
+
+				if err := newProvider.SubtractResourcesFor(consumer); err != nil {
+					return nil, []error{err}
+				}
+
+				providerDistribution.Providers = append(providerDistribution.Providers, newProvider)
+			}
+		}
+	}
+
+	p := findOptimalProviderDistribution(providerDistributions)
```

Comment:
ok, I see that multi errors + multiple scaling algorithms is throwing you for a loop a bit. When you encounter this, and you see your code start to become unwieldy, it means there's probably issues with the higher-level architecture. We need to first take a step back in scope (e.g. instead of "how do I improve calculateScaleUp?", it's "how do I improve EnvironmentScaler.Scale?"), and re-think how we want to handle all of these potential scenarios in the cleanest, simplest way possible.